lib/http/timeout/per_operation.rb

```diff
@@ -20,15 +20,15 @@ def initialize(*args)
       end
 
       def connect(socket_class, host, port, nodelay = false)
-        ::Timeout.timeout(@connect_timeout, TimeoutError) do
+        ::Timeout.timeout(@connect_timeout, ConnectTimeoutError) do
           @socket = socket_class.open(host, port)
           @socket.setsockopt(Socket::IPPROTO_TCP, Socket::TCP_NODELAY, 1) if nodelay
         end
       end
 
       def connect_ssl
-        rescue_readable(@connect_timeout) do
-          rescue_writable(@connect_timeout) do
+        rescue_readable(ConnectTimeoutError, @connect_timeout) do
```

Comment:
I'd argue that this should be part of write/read timeouts, and not connect. The operations under the hood for `connect_ssl` are read/write as part of the SSL handshake. It's not the same as a connect timeout because of DNS resolution or the server isn't accepting requests, so we shouldn't mask it that way.

(A nice side effect is it makes the code simpler, since don't have to pass classes to `rescue_####`)

Reply:
yeah, I went back and forth on this. Happy to change it to the way you suggest if that's what makes the most sense.

Reply:
I'd opt for consistency against the ops used personally, but @ixti or @tarcieri? Do either of you have any feelings on that.

Reply:
I think my reasoning for keeping it as-is would be to support the use case I described in #470. I mainly care about whether the request has been sent yet. As written this considers everything pre-request as a `ConnectionTimeoutError`, rather than giving read/write timeouts when the request hasn't even been sent (during the TLS handshake).

Reply:
[I have no strong feelings one way or the other](https://www.youtube.com/watch?v=CxK_nA2iVXw). But after couple of moments thinking and sketching I believe `ConnectTimeoutError` here is probably the best.

@zanker, as alternative we can make `connect_ssl` raise `HandshakeTimeoutError` that will be subclass of `ConnectTimeoutError` if you feel strong about that.

Reply:
Yeah, as much as I would like to simplify the implementation by using `ReadTimeoutError`, I also feel like `ConnectTimeoutError` would be more correct, so I'm ok with the current implementation.